Garbage collection tweaks and refactors.

MrsTonedOne opened the pull request:
We now handle the queue is nice little packaged layers.

Queue checking is allowed to finish before hard deletions begin (there would be cases where the queue would get reallllllly fucking big because hard deletes kept preventing it from clearing out the things that soft deleted.)

We store logging datums for use in tracking the stats of shit qdeleted

We now track how much time was spent hard deleting each type as a running total. This is printed into the qdel log.

Qdel now tracks how many of each type is ever qdeleted. this helps detect feedback loops of things that keep getting created and deleted for no reason.

Also added logging for destroy time as well as hard delete time, per type. As well as a few others

The processing of queues was improved so that it no longer forces byond to recreate the list for every item in the queue.

New qdel log window: 
![image](https://user-images.githubusercontent.com/7069733/30247394-83eccd48-95c8-11e7-81c6-28a9547b49ed.png)

@Cyberboss

MrsTonedOne committed:
Garbage collection queuing rewriting

optimumtact commented on code/controllers/subsystem/garbage.dm:
```diff
@@ -64,97 +70,178 @@ SUBSYSTEM_DEF(garbage)
+		var/value = D.vars[V]
+		switch(V)
+			if ("ckey")
+				D.vars[V] = null
```
won't this cause that logout bug you were talking about?

Cyberboss replied:
ghostize() should've been called in Destroy() it's not really the GC's issue if that happens

MrStonedOne replied:
I could record when this happens. it would likely nail down that one bug.

MrStonedOne replied:
Not like it matters, byond will likely be fixing that in 512, so 2 more years and it will be fixed

optimumtact commented on code/controllers/subsystem/garbage.dm:
```diff
@@ -64,97 +70,178 @@ SUBSYSTEM_DEF(garbage)
+//	ie: its the final attempt to fix the issue before just removing it.
+/datum/controller/subsystem/garbage/proc/EndodonticTherapy(datum/D)
+	var/static/list/exclude = list("locs", "color", "transform", "parent_type", "vars", "verbs", "type", "gc_destroyed")
+	for (var/V in D.vars - exclude)
```
hey that's a pretty good iea

optimumtact commented:
pretty good idea

optimumtact commented on code/controllers/subsystem/garbage.dm:
```diff
@@ -64,97 +70,178 @@ SUBSYSTEM_DEF(garbage)
+
+
+		if (islist(value))
+			value -= D
```
I don't get this one, if the var on D is a list, remove D from it?

RemieRichards replied:
MSO mentioned cyclic references being a common cause for GC fails.
I don't know of any object that DOES store itself in its own list, but this covers it anyway.

optimumtact replied:
ah I guess so, wouldn't it be sane to just cut/null the entire list

RemieRichards replied:
potentially but you might cause cascading deletes.
@MrStonedOne ?

Cyberboss replied:
Wouldn't that happen anyway were the object to GC normally?

Cyberboss replied:
Actually I take it back, the list could be owned by other things and `Cut()`ing it could severely fuck them

MrStonedOne replied:
I was going to null the variable, but in byond modifying certain builtin variables to null can cause oddities and i wanted to avoid adding too many to exempt or adding a second switch tree.

I suppose I could if it passes a byond special list check.

Cyberboss commented:
@MrStonedOne ?

MrStonedOne commented:
#30147 had delayed testing.

I'll rebase and get back on this.

MrsTonedOne committed:
Some Tweaks

MrStonedOne commented:
Still some oddities I'm working out.

But preliminary testing shows that out of the 5028 items that failed to GC and went thru the var purging, only 91 had to actually be deleted.

So very success.

optimumtact commented:
Is it possible we can check for it being in side a list on it  and report that case as a log?

I know it adds some extra special case handling to the switch, but it might be worth it

MrsTonedOne committed:
Fixes some queue tracking oddities

MrsTonedOne committed:
More tweaks

MrStonedOne commented:
done.

MrStonedOne commented:
This keeps breaking because its nulling static vars.

So I will have to either: come up with a way to detect what vars are static vars, or significantly underwhelm the usefulness of the feature by making it not null the things vars

MrStonedOne commented:
I'll re-work this one, maybe,

MrStonedOne commented:
Bringing this back baby.

I'm going to remove the clear vars thingy but this has some other changes that would be nice to have, such as tracking how much time was spent hard deleting objects.

MrsTonedOne committed:
Remove the middle two queue steps. tis was a silly idea

MrStonedOne commented:
This broke a TESTING feature, I need to fix it. I've half done that.

This is waiting on me finishing that.

MrsTonedOne committed:
New logging.

We store logging datums for use in tracking the stats of shit qdeleted

Added logging for destroy time as well as hard delete time, per type. As well as a few others

MrStonedOne commented:
@Cyberboss ready

MrsTonedOne committed:
Fix compile for testing

optimumtact commented:
![](https://file.house/Hm_T.png)

optimumtact commented:
did you open a byond feature request for being able to detect static vars?

optimumtact commented:
because it would be really nice to have it, and the only other way I can see doing it is prefixing all static vars with stat or static or somethign and then blacklisted that path.

AnturK commented:
What's the state on this, garbage is getting clogged almost every round it seems.

Cyberboss commented on code/controllers/subsystem/garbage.dm:
```diff
@@ -1,40 +1,44 @@
 	name = "Garbage"
 	priority = 15
-	wait = 20
-	flags = SS_POST_FIRE_TIMING|SS_BACKGROUND|SS_NO_INIT
```
PreInit runs on every subsystem, regardless of SS_NO_INIT, you can keep the flag

MrStonedOne replied:
I had at one point created an init proc, but ended up not not needing it. forgot to re-add no-init

Cyberboss commented on code/controllers/subsystem/garbage.dm:
```diff
@@ -46,116 +50,179 @@ SUBSYSTEM_DEF(garbage)
 	else
 		msg += "TGR:[round((totalgcs/(totaldels+totalgcs))*100, 0.01)]%"
+	msg += " P:[pass_counts.Join(",")]"
+	msg += "|F:[fail_counts.Join(",")]"
```
Should this string only be formatted in one place? I'm worried about a performance situation like we has with the blackbox

MrStonedOne replied:
yes... I should convert this to a list.

MrStonedOne replied:
But it would out of scope for this pr, as i'd have to convert every subsystem

Cyberboss commented on code/controllers/subsystem/garbage.dm:
```diff
@@ -46,116 +50,179 @@ SUBSYSTEM_DEF(garbage)
-	while(queue.len && starttime == world.time && starttimeofday == world.timeofday)
+/datum/controller/subsystem/garbage/proc/HandlePreQueue()
+	var/list/tobequeued = queues[GC_QUEUE_PREQUEUE]
+	var/static/count = 0
```
are we still worried about name collision with static vars? cause this one is pretty basic

MrStonedOne replied:
No, the person who mentioned that was a troll. they keep coming in with random usernames and pretending to know their shit.

Cyberboss commented on code/controllers/subsystem/garbage.dm:
```diff
@@ -46,116 +50,179 @@ SUBSYSTEM_DEF(garbage)
+	var/cut_off_time = world.time - collection_timeout[level] //ignore entries newer then this
+	var/list/queue = queues[level]
+	var/static/lastlevel
+	var/static/count = 0
```
and it's used twice...

Cyberboss commented on code/controllers/subsystem/garbage.dm:
```diff
@@ -166,40 +233,61 @@ SUBSYSTEM_DEF(garbage)
+	var/slept_destroy = 0	//Number of times it's slept in its destroy
+
+/datum/qdel_item/New(mytype)
+	name = mytype
```
you declared name as a string, but mytype is a type

MrsTonedOne committed:
Does the cyberboss

MrsTonedOne committed:
does the antur
